Functions/xASL_quant_HadamardDecoding.m

```diff
@@ -0,0 +1,165 @@
+function [imASL_reorder] = xASL_quant_HadamardDecoding(Encoded_ASL, xDecodingFields, NumberEchoTimes)
+%xASL_im_HadamardDecoding Hadamard-4 & Hadamard-8 Decoding
+%
+% FORMAT:       [Decoded_ASL] = xASL_im_HadamardDecoding(HadamardType, sec)
+%
+% INPUT:        HadamardType  - 4 for Hadamard4
+%                             - 8 for Hadamard8
+%                             - 12 for Hadamard12
+%               sec           - output from xASL_im_HadamardCaseByCase
+%
+%
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% DESCRIPTION:  Hadamard-4 & Hadamard-8 Decoding.
+%
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% EXAMPLE:      n/a
+% __________________________________
+% Copyright 2015-2021 ExploreASL
+
+% Decoding Fields
+
+TimeEncodedMatrixType = xDecodingFields.TimeEncodedMatrixType;
```

Comment:
What happens if this field doesn't exist? :)

Same for the other fields. You can check them, and give warnings if they are missing. This should of course not happen, but if it happens for whatever reason — e.g., there is something going wrong somewhere else in another function — then you are notified nicely, which can save you time in debugging later.

Also, an input field check is lacking, like:

`if margin<3 || isempty('variable name')`

Check other functions on how this should be done. Also, in the header you should note if an input is REQUIRED or OPTIONAL, and if OPTIONAL, what the default value is. See also in other functions.

Reply:
Valid point, we never really know if those exist if the data is a bit different. We could optimize this in a second PR where we also support the data of Koen though, because there those things will probably "break".

I'd suggest just to add some...

```matlab
if nargin ...
if isfield(...)
etc.
```

... for now and then we improve the logic within the Koen PR.